### PR TITLE
[DO NOT MERGE] Sample generate metrics rdl

### PR DESCRIPTION
**Pull request by OferLevi85**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSImpl.java [hunk 1]
```diff
@@ -8279,4 +8282,19 @@ public void run() {
...
+        String domainName = getRequestDomainName(ctx);
+        metric.increment("zms_api", domainName, principalDomainName, httpMethod, httpStatus, apiName);
+    }
+
+    public Object startMetricTiming() {
+        return metric.startTiming("zms_api_latency", null);
+    }
+
+    public void stopTiming(Object timerMetric, String httpMethod, int httpStatus, String apiName, ResourceContext ctx) {
+        String principalDomainName = getPrincipalDomain(ctx);
```
Comment (havetisyan) [inline servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSImpl.java:8296]: should make these 2 final as well.

File: servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSResources.java [hunk 2]
```diff
@@ -17,12 +17,15 @@
     @Path("/domain/{domain}")
     @Produces(MediaType.APPLICATION_JSON)
     public Domain getDomain(@PathParam("domain") String domain) {
+        int code = 200;
```
Comment (havetisyan) [inline servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSResources.java:20]: we should use ResourceException.OK

File: servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSResources.java [hunk 3]
```diff
@@ -17,12 +17,15 @@
...
+        ResourceContext context = this.delegate.newResourceContext(this.request, this.response);
+        Object timerMetric = this.delegate.startMetricTiming();
         try {
-            ResourceContext context = this.delegate.newResourceContext(this.request, this.response);
             context.authenticate();
-            return this.delegate.getDomain(context, domain);
+            Domain returnedValue = this.delegate.getDomain(context, domain);
```
Comment (havetisyan) [inline servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSResources.java:25]: not sure why this change was made?

File: servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSResources.java [hunk 4]
```diff
@@ -38,6 +41,9 @@ public Domain getDomain(@PathParam("domain") String domain) {
                 System.err.println("*** Warning: undeclared exception (" + code + ") for resource getDomain");
                 throw typedException(code, e, ResourceError.class);
             }
+        } finally {
+            this.delegate.stopTiming(timerMetric, "GET", code, "getDomain", context);
```
Comment (havetisyan) [inline servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSResources.java:45]: since the code is auto generated I'm guessing the generator needs to include the "getDomain" here and in the next line automatically. otherwise we can just set the method name in the resource context as well.

File: servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSResources.java [hunk 5]
```diff
@@ -17,12 +17,15 @@
     @Path("/domain/{domain}")
     @Produces(MediaType.APPLICATION_JSON)
     public Domain getDomain(@PathParam("domain") String domain) {
+        int code = 200;
+        ResourceContext context = this.delegate.newResourceContext(this.request, this.response);
+        Object timerMetric = this.delegate.startMetricTiming();
```
Comment (havetisyan) [inline servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSResources.java:22]: we should probably move this one line up since we should include the time required for context resource creation as well. Another option would be create the object within ResourceContext itself and then just get that on line 45 below.

File: servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSResources.java [hunk 6]
```diff
@@ -38,6 +41,9 @@ public Domain getDomain(@PathParam("domain") String domain) {
                 System.err.println("*** Warning: undeclared exception (" + code + ") for resource getDomain");
                 throw typedException(code, e, ResourceError.class);
             }
+        } finally {
+            this.delegate.stopTiming(timerMetric, "GET", code, "getDomain", context);
+            this.delegate.incrementMetric("GET", code, "getDomain", context);
```
Comment (havetisyan) [inline servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSResources.java:46]: Also see if you can combine these two calls into one generic update metric call and then internally we can make two separate calls - one for increment and one for the timer. at least the code generation might look simpler

File: servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSResources.java (hunk 4 above)
Comment (OferLevi85) [reply servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSResources.java:45]: Yes I assumed that while way we generate the function name in ZMSResources we can put it here

File: servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSResources.java (hunk 3 above)
Comment (OferLevi85) [reply servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSResources.java:25]: Mistake. I originally planned on stopping the timer just before returning the value (before figuring I can stop it in finally and have the error code if needed)

**Comment by OferLevi85** [comment]
Fixed it according to your comments so you can see the final result. No need to merge this of course. I've started working on the code generation.

File: servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSHandler.java [hunk 7]
```diff
@@ -96,4 +96,6 @@
     DomainRoleMembership getPendingDomainRoleMembersList(ResourceContext context, String principal);
     Schema getRdlSchema(ResourceContext context);
     ResourceContext newResourceContext(HttpServletRequest request, HttpServletResponse response);
+    void recordMetrics(String httpMethod, int httpStatus, String apiName, ResourceContext ctx);
```
Comment (havetisyan) [inline servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSHandler.java:99]: let's use the ResourceContext as the first argument to be consistent with other APIs

File: servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSHandler.java [hunk 8]
```diff
@@ -96,4 +96,6 @@
...
+    Object startMetricTiming();
```
Comment (havetisyan) [inline servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSHandler.java:100]: not sure why we need this? we should be able to hide the constructions of the metric within the newResourceContext api call so we don't need to expose this
